lib/modules/manager/deno/schema.ts

```diff
@@ -0,0 +1,429 @@
+  getSiblingFileName,
+  localPathIsFile,
+  readLocalFile,
+} from '../../../util/fs';
+import { regEx } from '../../../util/regex';
+import {
+  Json,
+  Jsonc,
+  LooseArray,
+  LooseRecord,
+} from '../../../util/schema-utils';
+import { joinUrlParts } from '../../../util/url';
+import { extractJsrPackageName } from '../../datasource/jsr/util';
+import { id as denoVersioningId, isValid } from '../../versioning/deno';
+import type { PackageDependency, PackageFile } from '../types';
+import type { DenoManagerData } from './types';
+import { denoLandRegex, depValueRegex } from './utils';
+
+// https://github.com/denoland/deno/blob/410c66ad0d1ce8a5a3b1a2f06c932fb66f25a3c6/cli/schemas/config-file.v1.json
+export interface DenoJsonFile {
+  name?: string;
+  version?: string;
+  lock?: string | boolean | { path?: string };
+  imports?: Record<string, string>;
+  scopes?: Record<string, Record<string, string>>;
+  importMap?: string;
+  tasks?: Record<string, string | { command?: string }>;
+  compilerOptions?: {
+    types?: string[];
+    jsxImportSource?: string;
+    jsxImportSourceTypes?: string;
+  };
+  lint?: {
+    plugins?: string[];
+  };
+  workspace?: string[] | { members: string[] };
+}
+
+export interface DepTypes
+  extends Omit<
+    Record<keyof DenoJsonFile, string>,
+    'name' | 'version' | 'workspace'
+  > {}
+
+export interface ImportMapJsonFile
+  extends Pick<DenoJsonFile, 'imports' | 'scopes'> {}
+
+export const DenoLock = Json.pipe(
+  // this schema is version 5
+  // https://github.com/denoland/vscode_deno/blob/7e125c6ffcdcdebd587f97be5341d404f5335b87/schemas/lockfile.schema.json
+  z.object({
+    version: z.string(),
+    specifiers: LooseRecord(z.string()).catch({}),
+    redirects: LooseRecord(z.string()).catch({}),
+    remote: LooseRecord(z.string()).catch({}),
+  }),
+).transform(({ version, specifiers, redirects, remote }) => {
+  const lockedVersions: Record<string, string> = {};
+  for (const [key, val] of Object.entries(specifiers)) {
+    // pick "7.1.3" from "7.1.3_pkgname@4.0.3_@types+pkgname@1.0.1"
+    const match = regEx(/^(?<lockedVersion>[^_\s]+)/).exec(val);
+    if (match?.groups?.lockedVersion) {
+      lockedVersions[key] = match?.groups?.lockedVersion;
+    }
+  }
+
+  const redirectVersions: Record<string, string> = {};
+  for (const [key, val] of Object.entries(redirects)) {
+    redirectVersions[key] = val;
+  }
+
+  const remoteVersions = new Set<string>();
+  for (const key of Object.keys(remote)) {
+    remoteVersions.add(key);
+  }
+  return {
+    lockedVersions,
+    redirectVersions,
+    remoteVersions,
+    lockfileVersion: Number(version),
+  };
+});
+
+export const Workspace = z
+  .union([
+    z.array(z.string()),
+    z.object({
+      members: z.array(z.string()),
+    }),
+  ])
+  .optional()
+  .transform((workspaces): string[] | undefined => {
+    if (!workspaces) {
+      return undefined;
+    }
+    return 'members' in workspaces ? workspaces.members : workspaces;
```

Comment:
Type checking shouldn't happen here, instead transform must be done at each branch of union, leading to the uniform result